internal/btf/co_re.go

```diff
@@ -0,0 +1,737 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return targetRes, nil
+}
+
+func coreParseSpec(typ Type, spec string, kind COREReloKind) (*coreSpec, error) {
+	if spec == "" || spec == ":" {
+		return nil, ErrInvalidCORESpec
+	}
+	var err error
+	coreSpec := coreSpec{
+		rootType: typ,
+		reloKind: kind,
+	}
+
+	if kind.isTypeBased() {
+		if spec == "0" {
+			return nil, ErrInvalidCORESpec
+		}
+		return &coreSpec, nil
+	}
+
+	rawIndexes := strings.SplitN(spec, ":", coreSpecMaxLen+1)
+	if len(rawIndexes) > coreSpecMaxLen {
+		return nil, fmt.Errorf("spec string too big")
+	}
+	if len(rawIndexes) == 0 {
+		return nil, ErrInvalidCORESpec
+	}
+	coreSpec.rawSpec = make([]uint32, len(rawIndexes))
+	for i, s := range rawIndexes {
+		idx, err := strconv.Atoi(s)
+		if err != nil {
+			return nil, fmt.Errorf("invalid spec index %s: %w", s, err)
+		}
+		coreSpec.rawSpec[i] = uint32(idx)
+	}
+
+	t := skipModsAndTypedefs(typ)
+
+	accessIdx := coreSpec.rawSpec[0]
+	acc := &coreAccessor{
+		typ: t,
+		idx: accessIdx,
+	}
+	coreSpec.spec = append(coreSpec.spec, acc)
+
+	if kind.isEnumValBased() {
+		e, ok := t.(*Enum)
+		if !ok || len(coreSpec.rawSpec) > 1 || accessIdx >= uint32(len(e.Values)) {
+			return nil, ErrInvalidCORESpec
+		}
+		acc.name = e.Values[accessIdx].Name
+		return &coreSpec, nil
+	}
+
+	if !kind.isFieldBased() {
+		return nil, ErrInvalidCORESpec
+	}
+
+	sz, err := Sizeof(t)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get size of type %v: %w", t, err)
+	}
+	coreSpec.bitOffset = accessIdx * uint32(sz) * 8
+
+	for i := 1; i < len(coreSpec.rawSpec); i++ {
+		t = skipModsAndTypedefs(t)
+		accessIdx = coreSpec.rawSpec[i]
+
+		if v, ok := t.(composite); ok {
+			m := v.members()[accessIdx]
+			coreSpec.bitOffset += m.Offset
+			if m.Name != "" {
+				coreSpec.spec = append(coreSpec.spec, &coreAccessor{
+					typ:  t,
+					idx:  accessIdx,
+					name: m.Name,
+				})
+			}
+			t = m.Type
+		} else if a, ok := t.(*Array); ok {
+			t = skipModsAndTypedefs(a)
+			flex := isFlexArray(coreSpec.spec[len(coreSpec.spec)-1], a)
+			if !flex && accessIdx >= a.Nelems {
+				return nil, fmt.Errorf("invalid array index")
+			}
+
+			coreSpec.spec = append(coreSpec.spec, &coreAccessor{
+				typ: t,
+				idx: accessIdx,
+			})
+			sz, err = Sizeof(t)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get size of type %v: %w", t, err)
+			}
+			coreSpec.bitOffset += accessIdx * uint32(sz) * 8
+		} else {
+			return nil, fmt.Errorf("relo for [%v] %s (at idx %d) captures type [%d] of unexpected kind %s", typ, spec, i, t.ID(), t)
+		}
+	}
+
+	return &coreSpec, nil
+}
+
+func findCandidates(localType Type, targetBtf *Spec) ([]Type, error) {
+	ln, ok := localType.(namer)
+	if !ok || ln.name() == "" {
+		return nil, fmt.Errorf("type %v must have a name", localType)
+	}
+	localEssenLen := Name(ln.name()).essentialNameLen()
+	var cands []Type
+
+	// index 0 is void
+	for i := 1; i < len(targetBtf.indexedTypes); i++ {
```

Comment:
Yes. This is where having access to the `btfKind` enum would be valuable, because then we wouldn't have to use `reflect.TypeOf` to check type equivalence.

Reply:
We need `indexedTypes` to create the initial relocation records: https://github.com/cilium/ebpf/pull/141/files#diff-b445609eec2b211de8e400221aec49e8R43

`Spec.FindType` also does not handle multiple candidate matches at the moment. This is possible with essential names since you can have multiple structs with the same essential name.

It does make sense to search first by essential name, and then do the type checking (since this is the expensive part). Perhaps we just need an additional map separate from `Spec.types` that is essentialName => candidates? `map[string][]Type`?

Reply:
> Spec.FindType also does not handle multiple candidate matches at the moment. This is possible with essential names since you can have multiple structs with the same essential name.

Yes, I'm not suggesting that you use FindType for CO-RE relocation, just that both share the same lookup map. I realised just now that this needs a little bit of extra logic to allow `FindType("thing___v52")` but that isn't too complicated I think.

> Perhaps we just need an additional map separate from Spec.types that is essentialName => candidates?

Also a possibility, but then we'd basically 2x the memory we use for lookup. I just checked, current vmlinux has something like 60k types. Maybe it doesn't matter.

Regardless of how we do it, essentialName lookup should be cached and then we can drop the double caching in calculateRelocation (I hope?)